Domain check

ElenaGrasovskaya opened the pull request:
Resolves https://github.com/Hypersequent/tms-issues/issues/453

Related PR https://github.com/Hypersequent/tms/pull/1254

![Screenshot from 2024-07-17 12-09-19](https://github.com/user-attachments/assets/75a4d69c-a1d1-489f-9d3c-0dd6af4af151)

[Screencast from 17.07.24 11:56:53.webm](https://github.com/user-attachments/assets/aa4591e0-1c71-4c66-b26e-9f43437f20c8)





<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
  - Added URL validity and reachability checks to JUnit upload functionality with error messaging for invalid URLs.

- **Tests**
  - Introduced test cases for the new `isUrlReachable` function to verify URL reachability through mocked DNS and HTTP requests.

- **Refactor**
  - Enhanced URL comparison logic in multiple test utilities to use pathname extraction for more precise URL matching.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ramilamparo commented:
Have you checked why the API calls are not failing if the domains are invalid in the first place?

ramilamparo commented on src/commands/junit-upload.ts:
```diff
@@ -35,11 +35,28 @@ export class JUnitUploadCommandModule implements CommandModule<unknown, JUnitArg
+		argv.middleware(async (args) => {
+			const runUrl = parseRunUrl(args);
+			if (!runUrl) {
+				console.warn(`Warning: The subdomain or run URL you've provided is invalid: ${args['run-url']}. Please provide a valid QA Sphere URL and try again.`);
```
The function throws an error with an invalid url, so this warning message wouldn't print at all.

ramilamparo commented on src/commands/junit-upload.ts:
```diff
@@ -35,11 +35,28 @@ export class JUnitUploadCommandModule implements CommandModule<unknown, JUnitArg
 		})
-
+		argv.showHelpOnFail(false)
+		argv.exitProcess(true)
```
isn't true the default?

ramilamparo commented on src/utils/misc.ts:
```diff
@@ -75,3 +79,48 @@ export const printError = (e: unknown) => {
+    return new Promise((resolve) => {
+      const requestModule = parsedUrl.protocol === 'https:' ? https : http;
+
+      const req = requestModule.request(
```
I think the `fetch` api would work just fine right?

AndrianBdn reviewed (changes requested):
Why are we checking for reachability at all?  

If somehow we have nothing to upload, we can report this to user: "no test cases, with QA Sphere test case IDs (P1-XXX) were found; no attempt to connect to domain.qasphere.com" 

Otherwise this checks are always making the tool to work slowly. Also it is not checking anything - there could be authorization error or some qasphere domain that does not exist like https://notexist.eu1.qasphere.com/project/TEST/run/32

ElenaGrasovskaya commented:
I changed the logic to sending the domain name as search param and added a small check under apiKey middleware on TMS. This should not slow down the tool.
